matrix-client/src/Network/Matrix/Internal.hs

```diff
@@ -156,8 +164,33 @@ instance FromJSON MatrixError where
 -- | 'MatrixIO' is a convenient type alias for server response
 type MatrixIO a = MatrixM IO a
 
-type MatrixM m a = m (Either MatrixError a)
+-- | The session record, use 'createSession' to create it.
+data ClientSession = ClientSession
+  { baseUrl :: Text,
+    token :: MatrixToken,
+    manager :: HTTP.Manager
+  }
 
+newtype MatrixM m a = MatrixM { unMatrixM :: ExceptT MatrixError (ReaderT ClientSession m) a }
+  deriving ( Functor
+           , Applicative
+           , Monad
+           , MonadError MatrixError
+           , MonadIO
+           , MonadThrow
+           , MonadCatch
+           , MonadMask
+           , MonadReader ClientSession
+           ) via (ExceptT MatrixError (ReaderT ClientSession m))
+
+instance MonadTrans MatrixM where
+  lift = MatrixM . lift . lift
+
+-- | Interpret MatrixM into your inner monad. Wraps the calls that
+-- interacts with the Matrix API.
+runMatrixM :: ClientSession -> MatrixM m a -> m (Either MatrixError a)
+runMatrixM = flip coerce
```

Comment:
That looks like magic :-)

Reply:
Haha, I can write out the constructors explicitly if you prefer. Its just:
```haskell
-- | Interpret MatrixM into your inner monad. Wraps the calls that
-- interacts with the Matrix API.
runMatrixM :: ClientSession -> MatrixM m a -> m (Either MatrixError a)
runMatrixM session = flip runReaderT session . runExceptT . unMatrixM
```

Reply:
Thanks, it's amazing ghc is happy with that single coerce :-)